File: SystemDesign/StrategyDesignPattern/sdp.py
"""
"Strategy Pattern lets you choose and combine interchangeable behaviors at runtime."
Robot example

There are three independent strategies.

Robot
 │
 ├── Walk Strategy
 ├── Talk Strategy
 └── Fly Strategy

Example 1:

Walk  → NormalWalk
Talk  → NormalTalk
Fly   → NoFly

Example 2:

Walk  → NoWalk
Talk  → NormalTalk
Fly   → NormalFly

Example 3:

Walk  → NormalWalk
Talk  → NoTalk
Fly   → NormalFly

You are composing the robot from different strategies.

"""


from abc import ABC, abstractmethod


# ---------------- Strategy Interface for Walk ----------------
class WalkableRobot(ABC):
    @abstractmethod
    def walk(self):
        pass


class NormalWalk(WalkableRobot):
    def walk(self):
        print("Walking normally...")


class NoWalk(WalkableRobot):
    def walk(self):
        print("Cannot walk.")


# ---------------- Strategy Interface for Talk ----------------
class TalkableRobot(ABC):
    @abstractmethod
    def talk(self):
        pass


class NormalTalk(TalkableRobot):
    def talk(self):
        print("Talking normally...")


class NoTalk(TalkableRobot):
    def talk(self):
        print("Cannot talk.")


# ---------------- Strategy Interface for Fly ----------------
class FlyableRobot(ABC):
    @abstractmethod
    def fly(self):
        pass


class NormalFly(FlyableRobot):
    def fly(self):
        print("Flying normally...")


class NoFly(FlyableRobot):
    def fly(self):
        print("Cannot fly.")


# ---------------- Robot Base Class ----------------
class Robot(ABC):
    def __init__(self, walk_behavior, talk_behavior, fly_behavior):
        self.walk_behavior = walk_behavior
        self.talk_behavior = talk_behavior
        self.fly_behavior = fly_behavior

    def walk(self):
        self.walk_behavior.walk()

    def talk(self):
        self.talk_behavior.talk()

    def fly(self):
        self.fly_behavior.fly()

    @abstractmethod
    def projection(self):
        pass


# ---------------- Concrete Robot Types ----------------
class CompanionRobot(Robot):
    def projection(self):
        print("Displaying friendly companion features...")


class WorkerRobot(Robot):
    def projection(self):
        print("Displaying worker efficiency stats...")


# ---------------- Main ----------------
if __name__ == "__main__":

    robot1 = CompanionRobot(
        NormalWalk(),
        NormalTalk(),
        NoFly()
    )

    robot1.walk()
    robot1.talk()
    robot1.fly()
    robot1.projection()

    print("-" * 30)

    robot2 = WorkerRobot(
        NoWalk(),
        NoTalk(),
        NormalFly()
    )

    robot2.walk()
    robot2.talk()
    robot2.fly()
    robot2.projection()

File: SystemDesign/StrategyDesignPattern/SDP.cpp
#include<bits/stdc++.h>
using namespace std;

class Talkable{
    public: 
        virtual void talk() = 0;
};
class NormalTalk : public Talkable{
public:
    void talk() override {
        cout << "Normal Talk" << endl;
    }
};
class NoTalk : public Talkable{
    public:
        void talk() override{
            cout << "No Talk" << endl;
        }
};

class Walkable{
    public:
        virtual void walk() = 0;
};
class NormalWalk : public Walkable{
    public:
        void walk() override{
            cout << "Normal Walk" << endl;
        }
};

class NoWalk : public Walkable{
    public:
        void walk() override{
            cout << "No Walk" << endl;
        }
};

class FlyAble{
    public:
        virtual void fly() = 0;
};
class NormalFly : public FlyAble{
    public:
        void fly() override{
            cout << "Normal Fly" << endl;
        }
};
class NoFly : public FlyAble{
    public: 
        void fly() override{
            cout << "No Fly" << endl;
        }
};

class Robot{
    public:
       Talkable* talkBehavior;
       Walkable* walkBehavior;
       FlyAble* flyBehavior;

       Robot(Talkable* tb, Walkable* wb, FlyAble* fb) {
           this->talkBehavior = tb;
           this->walkBehavior = wb;
           this->flyBehavior = fb;
       }

       void performTalk() {
           talkBehavior->talk();
       }

       void performWalk() {
           walkBehavior->walk();
       }

       void performFly() {
           flyBehavior->fly();
       }
};

class CompanionRobot : public Robot {
    public:
        CompanionRobot(Talkable* tb, Walkable* wb, FlyAble* fb) : Robot(tb, wb, fb) {}
};

int main(){
    CompanionRobot* robot = new CompanionRobot(new NormalTalk(), new NormalWalk(), new NormalFly());
    robot->performTalk();
    robot->performWalk();
    robot->performFly();
    delete robot;
    return 0;
}